toolkit/tools/graphanalytics/graphanalytics.go

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/traverse"
+	"gopkg.in/alecthomas/kingpin.v2"
+	"microsoft.com/pkggen/internal/exe"
+	"microsoft.com/pkggen/internal/logger"
+	"microsoft.com/pkggen/internal/pkggraph"
+	"microsoft.com/pkggen/internal/sliceutils"
+)
+
+const (
+	defaultMaxResults = "10"
+)
+
+// mapPair represents a key/value pair in a map[string][]string.
+// What the key and value represent are defined by the functions that use this.
+type mapPair struct {
+	key   string
+	value []string
+}
+
+var (
+	app            = kingpin.New("graphanalytics", "A tool to print analytics of a given dependency graph.")
+	inputGraphFile = exe.InputFlag(app, "Path to the DOT graph file to analyze.")
+	maxResults     = app.Flag("max-results", "The number of results to print per catagory. Set 0 to print unlimited.").Default(defaultMaxResults).Int()
+	logFile        = exe.LogFileFlag(app)
+	logLevel       = exe.LogLevelFlag(app)
+)
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	err := analyzeGraph(*inputGraphFile, *maxResults)
+	if err != nil {
+		logger.Log.Fatalf("Unable to analyze dependency graph, error: %s", err)
+	}
+}
+
+// analyzeGraph analyzes and prints various attributes of a graph file.
+func analyzeGraph(inputFile string, maxResults int) (err error) {
+	pkgGraph := pkggraph.NewPkgGraph()
+	err = pkggraph.ReadDOTGraphFile(pkgGraph, inputFile)
+	if err != nil {
+		return
+	}
+
+	printDirectlyMostUnresolved(pkgGraph, maxResults)
+	printDirectlyClosestToBeingUnblocked(pkgGraph, maxResults)
+
+	printIndirectlyMostUnresolved(pkgGraph, maxResults)
+	printIndirectlyClosestToBeingUnblocked(pkgGraph, maxResults)
+
+	return
+}
+
+// printIndirectlyMostUnresolved will print the top unresolved packages that are indirectly most blocking.
+func printIndirectlyMostUnresolved(pkgGraph *pkggraph.PkgGraph, maxResults int) {
+	unresolvedPackageDependents := make(map[string][]string)
+
+	for _, node := range pkgGraph.AllNodes() {
+		if node.Type != pkggraph.TypeRun && node.Type != pkggraph.TypeBuild {
+			continue
+		}
+
+		if node.State == pkggraph.StateUnresolved {
```

Comment:
>node.State == pkggraph.StateUnresolve [](start = 5, length = 37)

Why are we skipping unresolved nodes? Aren't these exactly what we wanted to print?

Reply:
This routine is generate a map of unresolved nodes to packages it blocks. Therefore we want to to traverse each real package (not unresolved) to find what is blocking it.

I'll add a comment to clarify.